k8s/negotiator.go

```diff
@@ -152,36 +154,101 @@ func (*KindNegotiatedSerializer) DecoderToVersion(d runtime.Decoder, _ runtime.G
 		}
+		if cast, ok := into.(resource.ListObject); ok {
+			logging.DefaultLogger.Debug("decoding object into provided resource.ListObject", "gvk", into.GetObjectKind().GroupVersionKind().String())
+			// TODO: use codec for each element in the list?
+			err := c.Decoder(data, cast)
+			return cast, defaults, err
+		}
 		if cast, ok := into.(*metav1.WatchEvent); ok {
+			logging.DefaultLogger.Debug("decoding object into provided *v1.WatchEvent", "gvk", into.GetObjectKind().GroupVersionKind().String())
 			err := c.Decoder(data, cast)
 			return cast, defaults, err
 		}
 		if cast, ok := into.(*metav1.List); ok {
+			logging.DefaultLogger.Debug("decoding object into provided *v1.List", "gvk", into.GetObjectKind().GroupVersionKind().String())
 			err := c.Decoder(data, cast)
 			return cast, defaults, err
 		}
-		// We shouldn't encounter this
-		// TODO: make better or return error?
+		if cast, ok := into.(*metav1.Status); ok {
+			logging.DefaultLogger.Debug("decoding object into provided *v1.Status", "gvk", into.GetObjectKind().GroupVersionKind().String())
+			err := c.Decoder(data, cast)
+			return cast, defaults, err
+		}
+
+		// TODO: This is the same process (just without casting) as WatchEvent, List, and Status (they all use the default Decoder). Should we still keep them separate?
+		logging.DefaultLogger.Debug("decoding object into provided unregistered resource using default Decoder", "gvk", into.GetObjectKind().GroupVersionKind().String())
 		err := c.Decoder(data, into)
 		return into, defaults, err
 	}
 
-	obj := c.SampleObject.Copy()
-	err := c.Codec.Read(bytes.NewReader(data), obj)
+	if defaults != nil {
+		if defaults.Kind == "Status" && defaults.Version == "v1" {
+			logging.DefaultLogger.Debug("decoding object into *v1.Status resource based on defaults", "gvk", defaults.String())
+			obj := &metav1.Status{}
+			err := c.Decoder(data, obj)
+			return obj, defaults, err
+		}
+		logging.DefaultLogger.Debug("defaults present", "gvk", defaults.String())
+	}
+
+	tm := indicator{}
+	err := c.Decoder(data, &tm)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error decoding object TypeMeta: %w", err)
+	}
+	if tm.GroupVersionKind().Version == "v1" && tm.GroupVersionKind().Kind == "Status" {
+		logging.DefaultLogger.Debug("decoding object into *v1.Status resource based on decoded TypeMeta", "gvk", tm.GroupVersionKind().String())
+		obj := &metav1.Status{}
+		err := c.Decoder(data, obj)
+		return obj, defaults, err
+	}
+	// Check if this is a List
+	if tm.Items != nil {
+		logging.DefaultLogger.Debug("decoding into a new empty list instance from kind", "gvk", tm.GroupVersionKind().String())
+		var obj resource.ListObject
+		if c.SampleList != nil {
+			obj = c.SampleList.Copy()
+		} else {
+			logging.DefaultLogger.Warn("no SampleObject set in CodecDecoder, using *resource.TypedList[*resource.UntypedObject]")
+			obj = &resource.TypedList[*resource.UntypedObject]{}
+		}
+		// TODO: use codec for each element in the list?
```

Comment:
This is my concern too – we are effectively decoding a list 2 times, which is extra CPU and memory.

I wonder if we can change `items` to be a `json.RawMessage` instead – that way we can keep current simple implementation (i.e. `err = c.Decoder(data, &obj)`) but avoid extra unmarshaling.

Reply:
I have added a `noAlloc` type that always acts as if it is correctly unmarshaled, but contains no data. Using a `*noAlloc` pointer for the `Items` field now, so it being non-nil can be used to check if the `items` key is present in the JSON payload.